[DOCS] Cloud docs for built-in actions (#11338)
great-expectations/great_expectations: docs/docusaurus/docs/cloud/alerts/trigger_actions.md

---
sidebar_label: 'Trigger actions'
title: 'Trigger actions'
description: Create and manage Actions based on the results of Validation runs.
---

import TabItem from '@theme/TabItem';
import Tabs from '@theme/Tabs';

import PrereqPythonInstalled from '../../core/_core_components/prerequisites/_python_installation.md';
import PrereqGxInstalled from '../../core/_core_components/prerequisites/_gx_installation.md';
import PrereqPreconfiguredDataContext from '../../core/_core_components/prerequisites/_preconfigured_data_context.md';
import PrereqValidationDefinition from '../../core/_core_components/prerequisites/_validation_definition.md';

> A Checkpoint executes one or more Validation Definitions and then performs a set of Actions based on the Validation Results each Validation Definition returns. This example will demonstrate how to create a `SlackNotificationAction`.

Reviewer: I see that this intro was largely copied from the Core "Create a Checkpoint with Actions" docs but I find it confusing in this context. One reason for this is that the first sentence includes multiple proper noun concepts the user might not be familiar with.  

Also, this intro is focused on the underlying technology rather than the value provided. (Lots of our docs intros are like this, it's something I've been trying to improve when the opportunity presents itself). A Checkpoint is a means to an end, not something the customer instinctively cares about in and of itself. 

I'd like to see the page intro lead with highlighting the value provided by Actions before getting into the details of how they work. The intro should help connect the problem the user knows they have to the solution we can provide. This should be done in plain language that will be understandable to new users and UI-first users. 

After the stage is set with the customer problem and plain language solution, the intro can then give an overview with branded language of the general workflow for implementing the solution. This will set customer expectations about level of effort and orient them to the overall process before they get into the detailed instructions or sample code. 

For inspiration, here are some examples of docs with roughly the kind of intro I'd like to see this page have

- [Manage alerts](https://docs.greatexpectations.io/docs/cloud/alerts/manage_alerts/) - leads with general value prop, then has bullet points detailing when the feature comes into play
- [Manage schedules](https://docs.greatexpectations.io/docs/cloud/schedules/manage_schedules/) - leads with general value prop, then explains how the feature works
- [Expectations overview](https://docs.greatexpectations.io/docs/cloud/expectations/expectations_overview/) this is a bit different because it's for conceptual content rather than instructional content. But I think it's a good example of defining a proper noun branded term with plain language instead of using other proper noun branded terms in the definition. 

Sometimes I like to look at competitor inspiration too - I think the first paragraph of https://docs.getmontecarlo.com/docs/notifications-v2 is nice, I like how it highlights specific third parties their notifications integrate with. That makes the page more findable through search for folks who know the tools their company uses that they want to integrate with GX but don't yet know the language we use for that (i.e. "Actions")

## Prerequisites

> - A [GX Cloud account](https://greatexpectations.io/cloud) with your Cloud access token and Cloud organization token saved in your environment variables

Reviewer: re: the phrasing of "Cloud access token and Cloud organization token" 
- since there are multiple types of access tokens (user and org) I think we should specify "user access token" for clarity and the best practice of using the option with the least privileges. 
- "Cloud organization token" sounds like a mashup of organization access token and organization id. We should use the term "organization ID" for clarity.

---

I think it would be helpful to link to instructions for getting and setting user access token and org id. Right now the best place for that is https://docs.greatexpectations.io/docs/cloud/connect/connect_python#get-your-user-access-token-and-organization-id and https://docs.greatexpectations.io/docs/cloud/connect/connect_python#set-the-gx-cloud-organization-id-and-user-access-token-as-environment-variables.

To avoid overloading one bullet point with too many links, it might be good to split this into one point for having the Cloud account and another for having credentials saved as env vars

> - A [GX Cloud account](https://greatexpectations.io/cloud) with your Cloud access token and Cloud organization token saved in your environment variables

Reviewer: the other prereqs all end in periods. this one should too for consistency

> - A Checkpoint.

Reviewer: Though we don't yet have instructions for creating Checkpoints that this can link to, I think it would be good to explicitly state here that the Checkpoint can be the GX-managed Checkpoint we automatically make for  GX-managed Expectations or a manually created Checkpoint the user set up for API-managed Expectations. 

I think being explicit about this would be especially helpful for UI-first users of GX-managed Expectations who likely have not come across the term "Checkpoint" before and don't know whether they need to do anything in particular to fulfill this prereq

- <PrereqPythonInstalled/>.

> - <PrereqGxInstalled/>.

Reviewer: These prereq snippets link to Core docs and use "Core" language. As part of reframing the relationship between Cloud and Core, I tried to eliminate all mentions of Core in the Cloud docs in favor of terms like Great Expectations Python package / GX library / GX Cloud API. And I tried to remove links from the Cloud docs to the Core docs as much as possible. 

The phrasing of the python prereq is ok. I suggest making the link go to https://www.python.org/downloads/ 

For the GX installation prereq, I suggest saying something like

"An installation of the Great Expectations Python library"
 
and linking to https://pypi.org/project/great-expectations/

## Procedure

<Tabs 
   queryString="procedure"
   defaultValue="instructions"
   values={[
      {value: 'instructions', label: 'Instructions'},
      {value: 'sample_code', label: 'Sample code'}
   ]}
>

> <TabItem value="instructions" label="Instructions">

Reviewer: The sample code tab includes some things not covered in the instructions tab. I think these should be included, especially since when I tried to follow the instructions the first time I got an error about SlackNotificationAction not being defined

```
import great_expectations as gx
from great_expectations.checkpoint import SlackNotificationAction

context = gx.get_context()
```

maybe `import great_expectations as gx` is duplicative of the prereq of having gx installed, but I thought installing and importing were separate things

> 1. Retrieve the Checkpoint to append the Action to.

Reviewer: I thought you mentioned you'd include a script for listing all Checkpoints to help folks retrieve the name of their GX-managed Checkpoint without guiding them to the UI / to smooth over some of the differences between working with GX-managed Checkpoints vs. API-managed Checkpoints. 

I still think that would be helpful to include, especially for folks working with GX-managed Checkpoints who likely don't yet know the name of their Checkpoint or how to find it. 

I'm not sure if there's a more targeted or performant way to get the info, but it looks like this will work

```
names = [checkpoint.name for checkpoint in context.checkpoints.all()]
print(names)
```

Author: I addressed this by instructing users to use the generate snippet function, since listing out the checkpoints isn't actually all that useful if they have multiple assets. What do you think?

Reviewer: +1 to listing all checkpoints being cumbersome if you have a lot of assets. I'm good with the approach you took. If we get customer questions about how to retrieve the names of API-managed Checkpoints, we can revisit the idea of including a script and see if it's possible to scope it to a single asset.

   ```python title="Python" name="docs/docusaurus/docs/cloud/alerts/_examples/create_a_checkpoint_with_actions.py - retrieve the Checkpoint"
   ```

> 1. Determine the Actions that the Checkpoint will automate.

Reviewer: I see that this language was copied from the core docs, but

- the word "automate" is getting kind of loaded with the various ways we use it when talking about expectation creation. I suggest using a different term here to avoid further overloading "automate". perhaps "trigger" to tie back into the name of the page
- the word "Determine" makes it sound like the user is picking from pre-configured options though the user has opportunity to control things like what channel a slack notification will go to. I suggest using a different word that speaks more the to configurability, perhaps "Define"

>    After a Checkpoint receives Validation Results from running a Validation Definition, it executes a list of Actions. The returned Validation Results determine what task is performed for each Action. Actions can include sending alerts when validations fail, or your own custom logic. The Actions list is executed once for each Validation Definition in a Checkpoint. The following is an example of how to append a `SlackNotificationAction` to the Action list of your Checkpoint.

Reviewer: This information feels out of place in this location.

This step is about defining the action. Background information about what the checkpoint does with validation results feels tangential to the discrete task the reader is trying to accomplish while following step 2. The last sentence mentions appending the action to the checkpoint but the instructions for that are in step 3, not step two.

Also this paragraph is somewhat duplicative of information already presented in the page intro. I suggest removing this paragraph in favor of having the intro provide an overview of how the feature works

>    After a Checkpoint receives Validation Results from running a Validation Definition, it executes a list of Actions. The returned Validation Results determine what task is performed for each Action. Actions can include sending alerts when validations fail, or your own custom logic. The Actions list is executed once for each Validation Definition in a Checkpoint. The following is an example of how to append a `SlackNotificationAction` to the Action list of your Checkpoint.

Reviewer: re: "your own custom logic" - I think we should leave this out for now since we won't have docs for Cloud custom actions in this first iteration.

>    Actions can be found in the `great_expectations.checkpoint` module. All Action class names end with `*Action`.

Reviewer: I think it would be nice to link to the relevant part of the API reference to help folks find what we're talking about here. unfortunately there's not a checkpoint landing page we can link to. maybe linking a nearby mention of SlackNotificationAction to https://deploy-preview-11338.docs.greatexpectations.io/docs/reference/api/checkpoint/SlackNotificationAction_class would be sufficiently helpful

   ```python title="Python" name="docs/docusaurus/docs/cloud/alerts/_examples/create_a_checkpoint_with_actions.py - create a SlackNotificationAction"
   ```

>    :::tip Setting up an EmailAction is separate from Email alerts controlled through the UI

Reviewer: I know I suggested including a tip like this but it feels a bit out of place while the page doesn't have an example of configuring an Email Action. I suggest removing it for now and we can add it back when we add an example of configuring an email action

>    Any `EmailActions` that are added to the list of Actions associated with a Checkpoint will activate separately from the Alerts controlled through the UI. For more information, see [Manage email alerts](docs/cloud/alerts/manage_email_alerts.md)

Reviewer: I suggested removing this tip, but if you decide to keep it, the last sentence should end with a period.

> 1. Append the Action to the Checkpoint Action list

Reviewer: internal nit: though this ordered list numbered with all `1`s in the markdown displays fine in the rendered page, we tend to use sequential numbers in the markdown. I prefer that for readability when editing files as a writer. When I have to update an existing step it makes it easier for me to find it. If a step refers to another by number it's easier to account for any changes that adding/removing steps imply if the numbers are sequential in the markdown.

>    Append the newly-created Action to the Checkpoint Action list and save the Checkpoint.

Reviewer: since there's no empty line before this, there's not a line break in the rendered page. This reads oddly.

Also the two lines feel redundant of each other

Can these be consolidated or disambiguated? 
<img width="676" height="234" alt="Screenshot 2025-08-27 at 12 48 41 PM" src="https://github.com/user-attachments/assets/d824331a-c13a-4c5c-bd69-77227a012af4" />

   ```python title="Python" name="docs/docusaurus/docs/cloud/alerts/_examples/create_a_checkpoint_with_actions.py - save the Checkpoint"
   ```

> 

Reviewer: I think it would be helpful to have some kind of what's next info / page conclusion after the user finishes working through the numbered instructions. Maybe an optional step number 4 where they run the Checkpoint to confirm the action is configured properly. Or maybe just acknowledging that the action will now be triggered when the Checkpoint is run. I'm not sure actually if scheduled validations will trigger actions appended to the GX managed Checkpoint or if those are only triggered when the user clicks the validate button in the UI or runs the Checkpoint with the API. 

Without some step to confirm the action is properly configured it seems possible that the user may proceed with a non-working action and then think no news is good news when they don't get notifications but in reality they could be experiencing false negatives.

---
sidebar_label: 'Trigger actions'
title: 'Trigger actions'
description: Create and manage Actions based on the results of Validation runs.
---

import TabItem from '@theme/TabItem';
import Tabs from '@theme/Tabs';

import PrereqPythonInstalled from '../../core/_core_components/prerequisites/_python_installation.md';
import PrereqGxInstalled from '../../core/_core_components/prerequisites/_gx_installation.md';
import PrereqPreconfiguredDataContext from '../../core/_core_components/prerequisites/_preconfigured_data_context.md';
import PrereqValidationDefinition from '../../core/_core_components/prerequisites/_validation_definition.md';

> Use Actions to notify the appropriate parties of the results of your Validation runs. These Actions can be triggered for either successful or failed validation runs. Validations are executed using Checkpoints, which each have a list of Actions that will be executed when each run has finished. By default, GX creates a Checkpoint for each Data Asset that you create. Optionally, you can also use a Checkpoint that you have created manually. This example will demonstrate how to create a `SlackNotificationAction` and append it to the list of Actions on a given Checkpoint.

Reviewer: suggestion to capitalize validation for consistency with first sentence

```suggestion
Use Actions to notify the appropriate parties of the results of your Validation runs. These Actions can be triggered for either successful or failed Validation runs. Validations are executed using Checkpoints, which each have a list of Actions that will be executed when each run has finished. By default, GX creates a Checkpoint for each Data Asset that you create. Optionally, you can also use a Checkpoint that you have created manually. This example will demonstrate how to create a `SlackNotificationAction` and append it to the list of Actions on a given Checkpoint.
```

> Use Actions to notify the appropriate parties of the results of your Validation runs. These Actions can be triggered for either successful or failed validation runs. Validations are executed using Checkpoints, which each have a list of Actions that will be executed when each run has finished. By default, GX creates a Checkpoint for each Data Asset that you create. Optionally, you can also use a Checkpoint that you have created manually. This example will demonstrate how to create a `SlackNotificationAction` and append it to the list of Actions on a given Checkpoint.

Reviewer: creating a GX-managed checkpoint is a Cloud-only thing so we should specify Cloud here

```suggestion
Use Actions to notify the appropriate parties of the results of your Validation runs. These Actions can be triggered for either successful or failed validation runs. Validations are executed using Checkpoints, which each have a list of Actions that will be executed when each run has finished. By default, GX Cloud creates a Checkpoint for each Data Asset that you create. Optionally, you can also use a Checkpoint that you have created manually. This example will demonstrate how to create a `SlackNotificationAction` and append it to the list of Actions on a given Checkpoint.
```

## Prerequisites

> - A [GX Cloud account](https://greatexpectations.io/cloud)

Reviewer: All items in this list should have consistent punctuation

```suggestion
- A [GX Cloud account](https://greatexpectations.io/cloud).
```

> - Your [Cloud access token and Cloud organization ID](../connect/connect_python#get-your-user-access-token-and-organization-id) saved in your environment variables.

Reviewer: This link sometimes gives a 404. Here's a video reproducing that behavior https://greatexpectationslabs.slack.com/archives/C03B8DZCJ07/p1756850226431309 

I think the following should work but please test loading it with and without trailing slash on the starting page to confirm

```suggestion
- Your [Cloud access token and Cloud organization ID](/cloud/connect/connect_python.md#get-your-user-access-token-and-organization-id) saved in your environment variables.
```

> - Your [Cloud access token and Cloud organization ID](../connect/connect_python#get-your-user-access-token-and-organization-id) saved in your environment variables.

Reviewer: Thanks for making most of the [changes I requested for this prereq](https://github.com/great-expectations/great_expectations/pull/11338#discussion_r2304579757). I still think it would be more clear to say "user access token" since there are  multiple types of access tokens. Do you think it's better for some reason to say more generically just "access token"?

```suggestion
- Your [Cloud user access token and Cloud organization ID](../connect/connect_python#get-your-user-access-token-and-organization-id) saved in your environment variables.
```

- A Checkpoint (either an automatically created GX-managed one or a manually created one).

> - [Python version 3.9 to 3.12](https://www.python.org/downloads/)

Reviewer: All items in this list should have consistent punctuation

```suggestion
- [Python version 3.9 to 3.12](https://www.python.org/downloads/).
```

> - [An installation of the Great Expectations Python library](https://pypi.org/project/great-expectations/)

Reviewer: All items in this list should have consistent punctuation

```suggestion
- [An installation of the Great Expectations Python library](https://pypi.org/project/great-expectations/).
```

## Procedure

<Tabs 
   queryString="procedure"
   defaultValue="instructions"
   values={[
      {value: 'instructions', label: 'Instructions'},
      {value: 'sample_code', label: 'Sample code'}
   ]}
>

<TabItem value="instructions" label="Instructions">
1. Import the relevant modules and instantiate your Context.
   ```python title="Python" name="docs/docusaurus/docs/cloud/alerts/_examples/create_a_checkpoint_with_actions.py - instantiate the Context"
   ```

2. Retrieve the Checkpoint to append the Action to.

   ```python title="Python" name="docs/docusaurus/docs/cloud/alerts/_examples/create_a_checkpoint_with_actions.py - retrieve the Checkpoint"
   ```

   :::tip The GX-managed Checkpoint name can be found through the UI

>    For the Data Asset of interest, go to the **Validations** tab, select the appropriate **Expectation Suite**, click the code snippet icon next to the **Validate** button, and then click **Generate snippet**.

Reviewer: I worry that "select the appropriate **Expectation Suite**" could be confusing to users working with an asset that only has the default GX-managed Expectation Suite. We don't show the **Expectation Suite** card in that case. I suggest making this step conditional

```suggestion
   For the Data Asset of interest, go to the **Validations** tab. If you have more than one **Expectation Suite**, select the **GX-managed** one. Then, click the code snippet icon next to the **Validate** button and click **Generate snippet**.
```

3. Define the Actions that the Checkpoint will trigger.
    

>    The following is an example of how to append a [`SlackNotificationAction`](../../reference/api/checkpoint/SlackNotificationAction_class) to the Action list of your Checkpoint.

Reviewer: This link is also giving inconsistent 404s. I think this should work but please test to confirm

```suggestion
   The following is an example of how to append a [`SlackNotificationAction`](/reference/api/checkpoint/SlackNotificationAction_class.mdx) to the Action list of your Checkpoint.
```

>    The following is an example of how to append a [`SlackNotificationAction`](../../reference/api/checkpoint/SlackNotificationAction_class) to the Action list of your Checkpoint.

Reviewer: appending happens in the next step. Here's a suggestion to make this description more specific to the code sample that directly follows

```suggestion
   The following is an example of how to define a [`SlackNotificationAction`](../../reference/api/checkpoint/SlackNotificationAction_class).
```

   ```python title="Python" name="docs/docusaurus/docs/cloud/alerts/_examples/create_a_checkpoint_with_actions.py - create a SlackNotificationAction"
   ```

4. Append the Action to the Checkpoint Action list.

>    Append the newly-created Action to the Checkpoint Action list and save the Checkpoint.

Reviewer: Thanks for fixing the display of these two lines. Even when displayed properly I still [think they feel a bit redundant](https://github.com/great-expectations/great_expectations/pull/11338/files#r2305152389) so I suggest dropping one for concision.

```suggestion
4.  Append the newly-created Action to the Checkpoint Action list and save the Checkpoint.
```

import PrereqPreconfiguredDataContext from '../../core/_core_components/prerequisites/_preconfigured_data_context.md';
import PrereqValidationDefinition from '../../core/_core_components/prerequisites/_validation_definition.md';

Use Actions to notify the appropriate parties of the results of your Validation runs. These Actions can be triggered for either successful or failed Validation runs. Validations are executed using Checkpoints, which each have a list of Actions that will be executed when each run has finished. By default, GX Cloud creates a Checkpoint for each Data Asset that you create. Optionally, you can also use a Checkpoint that you have created manually. This example will demonstrate how to create a `SlackNotificationAction` and append it to the list of Actions on a given Checkpoint.

## Prerequisites

> - A [GX Cloud account](https://greatexpectations.io/cloud)

Reviewer: This bullet point still needs ending punctuation to match the other items in the list

```suggestion
- A [GX Cloud account](https://greatexpectations.io/cloud).
```

> - Your [Cloud user access token and Cloud organization ID]/cloud/connect/connect_python.md#get-your-user-access-token-and-organization-id) saved in your [environment variables](/cloud/connect/connect_python#set-the-gx-cloud-organization-id-and-user-access-token-as-environment-variables).

Reviewer: The first link is missing an opening parenthesis. This makes it display as plain text in the rendered content rather than as a link

---

<img width="613" height="105" alt="Screenshot 2025-09-03 at 12 01 23 PM" src="https://github.com/user-attachments/assets/b061f7f5-b0c5-4a1f-99b3-cff03a9131ac" />

---

The most recent build failed. I think that's because the second link is missing `.md`

```suggestion
- Your [Cloud user access token and Cloud organization ID](/cloud/connect/connect_python.md#get-your-user-access-token-and-organization-id) saved in your [environment variables](/cloud/connect/connect_python.md#set-the-gx-cloud-organization-id-and-user-access-token-as-environment-variables).
```

   ```

   :::tip The GX-managed Checkpoint name can be found through the UI
   For the Data Asset of interest, go to the **Validations** tab. If you have more than one **Expectation Suite**, select the **GX-managed** one. Then, click the code snippet icon next to the **Validate** button, and then click **Generate snippet**.

3. Define the Actions that the Checkpoint will trigger.
    

>    The following is an example of how to define a [`SlackNotificationAction`](/reference/api/checkpoint/SlackNotificationAction_class.mdx) to the Action list of your Checkpoint.

Reviewer: Thanks for changing append to define. There's some lingering text that I suggested removing in the last round of feedback that I still think doesn't fit the scope of this step. I also don't think it makes sense with the changed verb. 

```suggestion
   The following is an example of how to define a [`SlackNotificationAction`](/reference/api/checkpoint/SlackNotificationAction_class.mdx).
```
